include/fiction/algorithms/physical_design/automatic_exhaustive_gate_designer.hpp

```diff
@@ -0,0 +1,276 @@
+                                                                     all_sidbs_in_cavas[cell_indices[j]]) < 0.5)
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    /**
+     * Add SiDB cells to a SiDB cell-level layout based on a vector of cell indices.
+     *
+     * This function creates a copy of the provided layout (`skeleton_layout`) and adds SiDB cells to it
+     * based on the cell indices provided in the `cell_indices` vector. If a cell at a specified index
+     * is of type `sidb_technology::cell_type::EMPTY`, it is replaced with `sidb_technology::cell_type::NORMAL`
+     * to represent the addition of SiDB cells.
+     *
+     * @param cell_indices A vector containing indices of cells in the layout to be added or modified.
+     * @return A copy of the original layout (`skeleton_layout`) with SiDB cells added at specified indices.
+     */
+    [[nodiscard]] Lyt add_cells_to_layout_based_on_indices(const std::vector<std::size_t>& cell_indices)
+    {
+        Lyt lyt_copy{skeleton_layout.clone()};
+
+        for (const auto i : cell_indices)
+        {
+            assert(i < all_sidbs_in_cavas.size() && "cell indices are out-of-range");
+            if (lyt_copy.get_cell_type(all_sidbs_in_cavas[i]) == sidb_technology::cell_type::EMPTY)
+            {
+                lyt_copy.assign_cell_type(all_sidbs_in_cavas[i], sidb_technology::cell_type::NORMAL);
+            }
+        }
+
+        return lyt_copy;
+    }
+};
+}  // namespace detail
+
+/**
+ * The *Automatic Exhaustive Gate Designer* designs SiDB gate implementations based on a specified Boolean
+ * function, a skeleton structure, canvas size, and a predetermined number of canvas SiDBs.
+ *
+ * It is composed of three steps:
+ * 1. In the initial step, all possible distributions of `number_of_sidbs` SiDBs within a given canvas are
+ * exhaustively determined. This ensures exhaustive coverage of every potential arrangement of ``number_of_sidbs`` SiDBs
+ * across the canvas.
+ * 2. The calculated SiDB distributions are then incorporated into the skeleton, resulting in the generation of distinct
+ * SiDB layouts.
+ * 3. The generated SiDB layouts then undergo an extensive simulation process. All input combinations possible for the
+ * given Boolean function are used to verify if the logic is fulfilled.
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ * @tparam TT The type of the truth table specifying the gate behavior.
+ * @param skeleton The skeleton layout used as a starting point for gate design.
+ * @param params Parameters for the *Automatic Exhaustive Gate Designer*.
+ * @return A vector of SiDB layouts that fulfill the given Boolean function (truth table).
+ */
+template <typename Lyt, typename TT>
+[[nodiscard]] std::vector<Lyt>
+automatic_exhaustive_gate_designer(const Lyt&                                           skeleton,
```

Comment:
I find it odd that a function is called `designer`. A function always **does** something, a class **is** something. I would have expected this to be a class when I read the name `designer`.

Reply:
Ah, good point. Thank you. I wasn't aware of this.